src/coreclr/jit/importercalls.cpp

```diff
@@ -32,6 +32,107 @@
 #pragma warning(disable : 21000) // Suppress PREFast warning about overly large function
 #endif
 
+bool Compiler::impTryFindField(CORINFO_METHOD_HANDLE methHnd, CORINFO_RESOLVED_TOKEN* pResolvedToken, OPCODE* opcode)
+{
+    // Either EE or JIT might throw exceptions below.
+    // If that happens, just don't inline the method.
+    //
+    struct Param
+    {
+        Compiler*              pThis;
+        CORINFO_METHOD_HANDLE  fncHandle;
+        CORINFO_CONTEXT_HANDLE exactContextHnd;
+        CORINFO_RESOLVED_TOKEN resolvedToken;
+        OPCODE                 opcode;
+        bool                   success;
+    } param;
+    memset(&param, 0, sizeof(param));
+
+    param.pThis           = this;
+    param.fncHandle       = methHnd;
+    param.exactContextHnd = MAKE_METHODCONTEXT(methHnd);
+    param.success         = false;
+
+    bool success = eeRunWithErrorTrap<Param>(
+        [](Param* pParam) {
+            // Cache some frequently accessed state.
+            //
+            Compiler* const       compiler    = pParam->pThis;
+            COMP_HANDLE           compCompHnd = compiler->info.compCompHnd;
+            CORINFO_METHOD_HANDLE ftn         = pParam->fncHandle;
+
+#ifdef DEBUG
+            if (JitConfig.JitNoInline())
+            {
+                pParam->success = false;
+                return;
+            }
+#endif
+
+            // Fetch method info. This may fail, if the method doesn't have IL.
+            //
+            CORINFO_METHOD_INFO methInfo;
+            if (!compCompHnd->getMethodInfo(pParam->fncHandle, &methInfo, pParam->exactContextHnd))
+            {
+                return;
+            }
+
+            // Speculatively check if initClass() can be done.
+            // If it can be done, we will try to inline the method.
+            CorInfoInitClassResult const initClassResult =
+                compCompHnd->initClass(nullptr /* field */, ftn /* method */, pParam->exactContextHnd /* context */);
+
+            if (initClassResult & CORINFO_INITCLASS_DONT_INLINE)
+            {
+                return;
+            }
+
+            // Given the VM the final say in whether to inline or not.
+            // This should be last since for verifiable code, this can be expensive
+            //
+            CorInfoInline const vmResult = compCompHnd->canInline(compiler->info.compMethodHnd, ftn);
+            if (vmResult == INLINE_FAIL || vmResult == INLINE_NEVER)
+            {
+                return;
+            }
+
+            pParam->resolvedToken.tokenContext = pParam->exactContextHnd;
+            pParam->resolvedToken.tokenScope   = methInfo.scope;
+            pParam->resolvedToken.tokenType    = CORINFO_TOKENKIND_Field;
+
+            auto code = methInfo.ILCode;
+            auto size = methInfo.ILCodeSize;
+            // instance getter
+            if (size == 7 && code[0] == CEE_LDARG_0 && code[1] == CEE_LDFLD && code[6] == CEE_RET)
```

Comment:
Can we move this check up so we can avoid redundant JIT-EE calls in the negative case? the checks looks cheaper than those

Reply:
Done